[merged] Rework libarchive import support

jlebon opened the pull request:
From the first commit message:

```
ostree-repo-libarchive.c: major refactor

- Make hardlink handling more generic. The previous strategy worked for
  tar archives, but not for cpio. It now works for both.
- Add support for SEL labeling (through the OstreeRepoCommitModifier)
- Add support for xattr_callback (through the OstreeRepoCommitModifier)
- Add support for filter (through the OstreeRepoCommitModifier)
- Add a use_ostree_convention option
```

The remaining commits add tests for the new features.

---

The main motivation for this work is https://github.com/projectatomic/rpm-ostree/pull/107. With these patches, `rpm-ostree` will be able to much more easily import RPMs into repos, with full support for filecaps and SELinux labeling.

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
+path_relative (const char *src)
+{
+  if (src[0] == '.' && src[1] == '/')
+    src += 2;
```
Doesn't this also need to be a loop so we canonicalize paths like `././foo` ?

jlebon replied:
Good point.

jlebon commented:
:arrow_up: Fixup added!

cgwalters commented:
Only took a quick scan, but the high level looks really good.  Will look more next week.

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -50,10 +50,10 @@ propagate_libarchive_error (GError      **error,
-  else if (src[0] == '.' && src[1] == '\0')
-    src++;
   while (src[0] == '/')
     src++;
```
This still won't handle `/./foo` right?  I think the code needs to be more like:

```
while (TRUE)
  {
    if (src[0] == '.' && src[1] == '/')
      {
        src += 2;
        continue;
      }
   if (src[0] == '/')
     {
       src++;
       continue
     }
   break;
}
```

Or something?  There's got to be some LGPL-or-below code we can look at as a reference...okay, some quick `git grep relative` in libarchive led me to [strip_absolute_path](https://github.com/libarchive/libarchive/blob/e61afbd463d1b06fcf0d7c0e37feeb9334954af6/tar/util.c#L398) which is rather terrifying...we could likely drop the Windows code at least.

jlebon replied:
Ahhh OK, I will tweak this so we at least handle the non-expensive cases.

cgwalters commented on tests/test-libarchive-import.c:
```diff
@@ -243,6 +301,166 @@ test_libarchive_ignore_device_file (gconstpointer data)
+                const char  *path,
+                GFileInfo   *file_info,
+                gpointer     user_data)
+  {
```
clang doesn't support GNU C nested functions, and I'm kind of with them there.  Instead it implements [blocks](https://en.wikipedia.org/wiki/Blocks_%28C_language_extension%29).

cgwalters replied:
(To clarify I'm suggesting we use neither for now, i.e. just hoist this out)

jlebon replied:
Sounds good.

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
+  if (g_str_has_suffix (final, "/"))
+    {
+      g_autofree char *tmp = final;
+      final = g_strndup (final, strlen (final) - 1);
```
Also need to handle multiple trailing `/` here?  Would be a bit more efficient too to get the end pointer, then walk backwards and replace them with `NUL` in place.

jlebon replied:
Smart! I like that.

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
+        return FALSE;
+
+      if (xattrs)
+        g_variant_unref (xattrs);
```
Doesn't this need to be above the `return`?

jlebon replied:
Hmm, I don't think so. If `builder_add_label()` fails, the xattrs will be cleaned up by the cleanup func. If `builder_add_label()` succeeds, we need to drop the ref we have to the user's xattrs, if any, so that we can replace it with the variant ref we get from our builder (which already has the xattrs the user returned + the selabel).

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
+  GSList *slist;
+
+  slist = g_hash_table_lookup (ctx->deferred_hardlinks, hardlink);
+  new_slist = slist == NULL;
```
IMO parses more easily visually as: `new_slist = (slist == NULL)`

jlebon replied:
Will do.

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
+    if (!aic_import_file (ctx, parent, path, fi, cancellable, error))
+      return FALSE;
+
   if (hardlink)
```
This could be `else if (hardlink)` right?

jlebon replied:
We need to also add to the list hardlinks which carry a payload so that they can be found later. E.g. in `aic_import_from_hardlink` we correct the initial non-hardlink file to point to the hardlink csum.

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
-#endif
+
+static void
+slist_free (void *data)
```
I'd call this `deferred_hardlinks_free` or so.

jlebon replied:
Sounds good.

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
+    g_object_unref (dh->parent);
+    g_free (dh->path);
+    g_slice_free (DeferredHardlink, dh);
+  }
```
Another nested function.

jlebon replied:
Will pull out.

cgwalters commented on src/ostree/ot-builtin-commit.c:
```diff
@@ -147,13 +149,68 @@ parse_statoverride_file (GHashTable   **out_mode_add,
+    {
+      const char *line = *iter;
+
+      if (*line == '-')
```
My memory is a bit dim on why the statoverride files start with `+`...maybe so they could have comments with `#`?  Dunno.  If so I doubt it was used.

It seems like it'd be simpler to just have a newline-separated list.  (Anyone who complains about this should really be using the API IMO)

jlebon replied:
Sure, I'll rework stat_override and this to be simpler and more memory efficient.

cgwalters replied:
Well I don't think we can change the statoverride format, it'd break gcontinuous (and maybe other people have discovered it)

jlebon replied:
The leading + will still work actually. It goes through `g_strtod()` which allows it. It's just that we don't require it now.

cgwalters commented on src/ostree/ot-builtin-commit.c:
```diff
@@ -147,13 +149,68 @@ parse_statoverride_file (GHashTable   **out_mode_add,
+    goto out;
+
+  ret_hash = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  lines = g_strsplit (contents, "\n", -1);
```
Could avoid lots of mallocs with iteration via `strchr (contents, '\n')`...doesn't matter for this case but just pointing it out as something to keep in mind for other more perf sensitive code.

jlebon replied:
I started working on doing this but felt silly doing it, so I went back to using `g_strsplit()`. It just felt like the trade-off in legibility was not worth it for something that people should probably use the API for if it somehow becomes a bottleneck.

cgwalters replied:
Yeah sorry, I meant to make clear I wasn't saying we shouldn't change this code, agree it's not worth it.

jlebon replied:
Right, makes sense now rereading your comment. :+1: I do like to "do it right" usually even when it's for a trivial case. But yeah, guess it wasn't worth it here.

jlebon commented:
Added fixups! :arrow_up:

cgwalters commented on src/ostree/ot-builtin-commit.c:
```diff
@@ -97,104 +97,64 @@ static GOptionEntry options[] = {
-        }
-    }
+  lines = g_strsplit (contents, "\n", -1);
+  for (char **iter = lines; iter && *iter; iter++)
```
I'd be a bit more comfortable if we had braces `{` for nested loops/conditionals...otherwise it's too easy down the line to add an additional statement and have it go in the wrong section.

cgwalters replied:
(Although https://gnu.wildebeest.org/blog/mjw/2016/01/09/looking-forward-to-gcc6-nice-new-warnings/ will help with that)

cgwalters commented on src/ostree/ot-builtin-commit.c:
```diff
@@ -97,104 +97,64 @@ static GOptionEntry options[] = {
+  lines = g_strsplit (contents, "\n", -1);
+  for (char **iter = lines; iter && *iter; iter++)
+    if (**iter != '\0') /* skip empty lines at least */
+      if (!cb (*iter, cbdata, error))
```
Or alternatively here combine.

cgwalters commented on tests/basic-test.sh:
```diff
@@ -181,15 +181,15 @@ echo "ok commit with no xattrs"
 cd ${test_tmpdir}
 cat > test-statoverride.txt <<EOF
-+2048 /a/nested/3
+2048 /a/nested/3
```
Can we add a new test, since we still need `+` to work?

jlebon commented:
I rebased and added fixups to address the last comments. I squashed all the previous fixups and only kept the new ones.

Since for now we're punting on the name:id issues some clients will face, I went a different way than we had originally talked about re. passing hashtables through the `OstreeRepoImportArchiveOptions`. I simply added a `callback_with_entry_pathname` flag which causes the path sent to the filter & xattr callbacks to match the original entry from the archive. This will be useful for e.g. `rpm-ostree` so that the unpacker can match the entry with rpmfi overrides.

For the name:id issues, we do not impose any restrictions in the ImportArchive API. Instead, client apps like `rpm-ostree` should do the checking themselves (e.g. in the filter callback) and report an error to the user if it does not yet support uid/gid != 0 (see also https://github.com/projectatomic/rpm-ostree/issues/49).

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -45,269 +47,715 @@ propagate_libarchive_error (GError      **error,
+   * archive. For us however, this will be important. So we do our best to deal
+   * with non-conventional paths (but note that we don't try to be sophisticated
+   * enough to handle e.g. ../ & ./ in the middle of normal dir comps, which
+   * would probably require a stack). Also important, we relativize the path. */
```
Note `ot_util_path_split_validate()` *does* intentionally explicitly error out if we find `..` as a path component.  See also `ot_util_filename_validate()`.

I think with the current logic it'd be possible to have `foo/../../../../bar`...and we wouldn't trip any assertions =/

We really need unit tests for this function.

I submitted https://github.com/ostreedev/ostree/pull/283

jlebon commented:
:arrow_up: Added a fixup to validate all paths using `ot_util_path_split_validate()`.

cgwalters commented on src/libostree/ostree-repo-libarchive.c:
```diff
@@ -81,13 +82,23 @@ path_relative (const char *src)
     src += 1;
 
+  /* make sure that the final path is valid (no . or ..) */
+  if (!ot_util_path_split_validate (src, NULL, error))
```
A bit expensive to do it this way, we end up mallocing all the path components just to throw them away.  But eh...we can revisit this later.

cgwalters commented:
@cgwalters-bot r+ 37c09a3
